sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala

```diff
@@ -1184,6 +1186,137 @@ case class ArraySort(child: Expression) extends UnaryExpression with ArraySortLi
   override def prettyName: String = "array_sort"
 }
 
+/**
+ * Returns a random permutation of the given array.
+ *
+ * This implementation uses the modern version of Fisher-Yates algorithm.
+ * Reference: https://en.wikipedia.org/wiki/Fisher%E2%80%93Yates_shuffle#Modern_method
```

Comment:
is it safe to use Fisher-Yates here? AFAIK we should not change the input value, in case it's used by other expressions.

Reply:
I copied the input array before starting shuffle not to change the input value. Isn't it safe?

Reply:
if we create a new array, I guess there should be some simpler algorithms without swapping...

Reply:
Oh, I see. Let me try.

Reply:
In the latest commit, "inside-out" looks simpler without swapping (using just an assignment).